### PR TITLE
Oregon Scientific V1 Decoding

### PR DESCRIPTION
**Pull request by ygator**

### CODE REVIEW

File: src/devices/oregon_scientific_v1.c [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+			int    sid      = nibble[0];
+			int    channel  = ((nibble[1] >> 2) & 0x03) + 1;
+			int    uk1      = (nibble[1] >> 0) & 0x03;	/* unknown.  Seen change every 60 minutes */
+			float  tempC    =  nibble[2] / 10. + nibble[3] + nibble[4] * 10.;
+			int    battery  = (nibble[5] >> 3) & 0x01;
+			int    uk2      = (nibble[5] >> 2) & 0x01;	/* unknown.  Always zero? */
+			int    sign     = (nibble[5] >> 1) & 0x01;
+			int    uk3      = (nibble[5] >> 0) & 0x01;	/* unknown.  Always zero? */
+			int    checksum = nibble[6] + 16 * nibble[7];
+
```
Comment (merbanan) [inline src/devices/oregon_scientific_v1.c]: 16 * nibble[7] change that to the bit shift it really is.

File: src/devices/oregon_scientific_v1.c [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+int ret = 0;
+char time_str[LOCAL_TIME_BUFLEN];
+int i;
+int nibble[OSV1_BITS/4];
+int cs;
+uint32_t uRaw;
+data_t *data;
+static time_t lastTime = 0;
+static uint32_t lastuRaw;
+
```
Comment (merbanan) [inline src/devices/oregon_scientific_v1.c]: Indent this

File: src/devices/oregon_scientific_v1.c [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+			int    uk3      = (nibble[5] >> 0) & 0x01;	/* unknown.  Always zero? */
+			int    checksum = nibble[6] + 16 * nibble[7];
+
+			if(sign) tempC = -tempC;
+			float tempF = tempC * 1.8 + 32;
+
+			// printf("OSv1 %s  %s  sid %1x  channel %d  uk1 %1x  tempC %3.1f  battery %d  uk2 %d  sign %d  uk3 %d  checksum %02x  cs %02x\n",
+			//	time_str, raw, sid, channel, uk1, tempC, battery, uk2, sign, uk3, checksum, cs);
+			if(checksum == cs) {
+				/* supress duplicate message */
```
Comment (merbanan) [inline src/devices/oregon_scientific_v1.c]: Move the checksum check earlier

File: src/pulse_demod.c [hunk 4]
```diff
@@ -319,3 +319,68 @@ int pulse_demod_clock_bits(const pulse_data_t *pulses, struct protocol_state *de
 
    return events;
 }
+
+int pulse_demod_osv1(const pulse_data_t *pulses, struct protocol_state *device) {
+	int events = 0;
```
Comment (merbanan) [inline src/pulse_demod.c]: Can you add a small description how this modulation works ?

**Comment by merbanan** [comment]
Squash the commits and fix the comments and the patch is fine.

File: src/devices/oregon_scientific_v1.c [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+			float  tempC    =  nibble[2] / 10. + nibble[3] + nibble[4] * 10.;
+			int    battery  = (nibble[5] >> 3) & 0x01;
+			int    uk2      = (nibble[5] >> 2) & 0x01;	/* unknown.  Always zero? */
+			int    sign     = (nibble[5] >> 1) & 0x01;
+			int    uk3      = (nibble[5] >> 0) & 0x01;	/* unknown.  Always zero? */
+			int    checksum = nibble[6] + 16 * nibble[7];
+
+			if(sign) tempC = -tempC;
+			float tempF = tempC * 1.8 + 32;
+
```
Comment (merbanan) [inline src/devices/oregon_scientific_v1.c]: tempF is unused.

File: src/devices/oregon_scientific_v1.c (hunk 1 above)
Comment (merbanan) [inline src/devices/oregon_scientific_v1.c]: Declare all the variables in the top.

File: src/devices/oregon_scientific_v1.c [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+			// printf("OSv1 %s  %s  sid %1x  channel %d  uk1 %1x  tempC %3.1f  battery %d  uk2 %d  sign %d  uk3 %d  checksum %02x  cs %02x\n",
+			//	time_str, raw, sid, channel, uk1, tempC, battery, uk2, sign, uk3, checksum, cs);
+			if(checksum == cs) {
+				/* supress duplicate message */
+				if((time(NULL) - lastTime) >= 5 || uRaw != lastuRaw) {
```
Comment (merbanan) [inline src/devices/oregon_scientific_v1.c]: How often does it send duplicate messages ?

**Comment by merbanan** [comment]
Did you upload sample signals ?

**Comment by ygator** [comment]
Yes there are signals.
Commits squashed.  Changes made.

**Pull request merged**